doc/coding.md

```diff
@@ -75,3 +78,77 @@ import (
   or retry for it and/or is fully recoverable.
 * Use log level `ERROR` when something occurs which is fatal to the operation,
   but not to the service or application.
+
+### Use of goto
+
+Use of `goto` is generally frowned up on in higher level languages. We use
+`goto` statements for the following specific uses:
+* Ensure RPCs always return a reply to caller
+* Getting out of nested loops
+* Auto-generated code
+
+Please use `defer()` for ensuring that relevant resource cleanups happen when a
+function/method exits. Also use `defer()` to revert something on a later
+failure.
+
+Developers with significant experience in C should be careful not to
+excessively use `goto` just to ensure single exit point to a function. Unlike
+C programs, there is no memory to be free()d here. Care must be taken when one
+ports code from glusterd1 (c) to glusterd2 (go).
+
+### glusterd2 specific conventions
+
+**Do not log at the caller of `txn.*` methods:**
+
+Certain patterns repeat very often in codebase. For reducing clutter, we have
+moved some of the logging at the caller to inside the function. One such
+instance are the methods of `transaction.Context` interface, which are used at
+so many places. They log internally and caller shouldn't be logging. For
+example:
+
+```go
+if err := txn.Ctx.Set("req", &req); err != nil {
+        // do NOT log here
+        restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+        return
+}
+```
+
+**Use log.WithError() to log errors**
+
+It is common pattern to log the error received as a field in the log entry.
+Please use `WithError` for the same:
+
+Do this:
+```go
+log.WithError(err).WithField("path", path).Error("Failed to delete path")
+```
+
+Do NOT do this:
+```go
+        log.WithFields(log.Fields{
+                "error": err,
+                "path":  path,
+        }).Error("Failed to delete path")
+```
+
+**Reusing transaction steps**
+
+If a transaction step is used in more than one transactions, re-register the
```

Comment:
Well, the entire "glusterd2 specific conventions" section needs some glusterd2 specific context. I'll try to rephrase it and rephrase the PR :+1: 

If you have a specific way to describe this, I'm all ears too :)

Reply:
Once registered by the original user of the step, there is no need to re-register steps again. You can just use them registered name when building the new-txn.

Reply:
@kshlm If not re-registering, the naming can be more clearer. I've seen pattern such as this before: using "vol-option.StoreVolume" using in operations other than setting volume option.

On a second thought, this need not be in coding convention doc as it may be too specific.